server/src/main/java/io/crate/replication/logical/action/TransportDropSubscriptionAction.java

```diff
@@ -21,130 +21,283 @@
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    logger.error("Couldn't execute task 'drop-sub-close-tables'. Please run command DROP SUBSCRIPTION again.");
+                    future.completeExceptionally(e);
+                }
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    future.complete(null);
+                }
+            }
+        );
+        return future;
+    }
+
+    /**
+     * Removes subscription from the cluster metadata and
+     * removes setting LogicalReplicationSettings.REPLICATION_SUBSCRIPTION_NAME from each relevant for subscription index.
+     */
+    private CompletableFuture<Void> submitDropSubscriptionTask(String subscriptionName,
+                                                               boolean ifExists,
+                                                               List<IndexMetadata> subscriptionIndices) {
+        var future = new CompletableFuture<Void>();
+
+        // Check if we're still the elected master before sending second update task
+        if (!clusterService.state().nodes().isLocalNodeElectedMaster()) {
+            logger.error("Couldn't execute task 'drop-sub-remove-setting-and-metadata'. Please run command DROP SUBSCRIPTION again.");
+            future.completeExceptionally(new IllegalStateException("Master was re-elected, cannot execute 'drop-sub-remove-setting-and-metadata'"));
+        }
+
+        clusterService.submitStateUpdateTask(
+            "drop-sub-remove-setting-and-metadata",
+            new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    Metadata currentMetadata = currentState.metadata();
+                    Metadata.Builder mdBuilder = Metadata.builder(currentMetadata);
+
+                    SubscriptionsMetadata oldMetadata = (SubscriptionsMetadata) mdBuilder.getCustom(SubscriptionsMetadata.TYPE);
+                    if (oldMetadata != null && oldMetadata.subscription().containsKey(subscriptionName)) {
+
+                        SubscriptionsMetadata newMetadata = SubscriptionsMetadata.newInstance(oldMetadata);
+                        newMetadata.subscription().remove(subscriptionName);
+                        assert !newMetadata.equals(oldMetadata) : "must not be equal to guarantee the cluster change action";
+                        mdBuilder.putCustom(SubscriptionsMetadata.TYPE, newMetadata);
+
+                        // mdBuilder is mutated accordingly.
+                        removeSubscriptionSetting(subscriptionIndices, mdBuilder);
+
+                        return ClusterState.builder(currentState).metadata(mdBuilder).build();
+                    } else if (ifExists == false) {
+                        throw new SubscriptionUnknownException(subscriptionName);
```

Comment:
Just to bring attention - initially wanted to log and completeExceptionally but II think it actually goes to onFailure so no need to do that.